Sources/HTTP/HTTPServer.swift

```diff
@@ -1,54 +1,336 @@
-// This source file is part of the Swift.org Server APIs open source project
 //
-// Copyright (c) 2017 Swift Server API project authors
-// Licensed under Apache License v2.0 with Runtime Library Exception
+//  HTTPServer.swift
+//  HTTP
 //
-// See http://swift.org/LICENSE.txt for license information
+//  Created by Helge Hess on 22.10.17.
+//  Copyright © 2017 ZeeZide GmbH. All rights reserved.
 //
 
-/// Definition of an HTTP server.
-public protocol HTTPServing: class {
+import Foundation
+import Dispatch
+#if os(Linux)
+  import Glibc
+#else
+  import Darwin
+#endif
 
-    /// Start the HTTP server on the given `port`, using `handler` to process incoming requests
-    func start(port: Int, handler: @escaping HTTPRequestHandler) throws
+public class HTTPServer {
+  
+  public enum SocketError : Swift.Error {
+    case setupFailed      (Int32)
+    case couldNotSetOption(Int32)
+    case bindFailed       (Int32)
+    case listenFailed     (Int32)
+  }
+  public enum Error : Swift.Error {
+    // TBD: this could be done in a better way
+    case socketError(SocketError)
+  }
+  
+  private  var boundAddress   : sockaddr_in? = nil
+  
+  private  var listenSource   : DispatchSourceRead?
+  
+  internal let queue =
+                 DispatchQueue(label: "de.zeezide.swift.server.http.server")
+  private  var connections = [ HTTPConnection ]()
+  
+  private  var acceptCount       = 0
+  private  let handlerBaseQueues : [ DispatchQueue ] = {
+    var queues = [ DispatchQueue ]()
+    for i in 0..<ProcessInfo.processInfo.processorCount {
+      let q = DispatchQueue(label: "de.zeezide.swift.server.http.handler\(i)")
```

Comment:
Interesting - the idea here is that its most efficient to have a queue per processor?

Reply:
That is not an idea, that is a fact ;-) There is also `activeProcessCount` which might be more appropriate (is that ever different?)

All this really belongs into the Options, care to file an Issue for this? E.g. the above is good for a server, but an on-device server may only want to use one thread.

Reply:
Hm cool didn't know about that

```
Whereas the processorCount property reports the number of advertised processing cores, the activeProcessorCount property reflects the actual number of active processing cores on the system. There are a number of different factors that may cause a core to not be active, including boot arguments, thermal throttling, or a manufacturing defect.
```
I guess they can be different

Reply:
Shouldn't this be the responsibility of `libdispatch`? Having 1 warm thread per processor is great, but GCD is (partly among other things) supposed to free us from having to worry about this detail (# of threads per process, and arguable # of active threads at all).

If we think it'll be commonly desirable to warm threads for this kind of application (a server written in Swift), can we move this logic behind a GCD API or have GCD (if it doesn't do this already) warm threads in a library constructor function? At the very least, I think we should open a ticket/issue and leave a TODO here to come back and clean this up.

Reply:
`DispatchQueue.concurrentPerform` already leans on the internal function `_dispatch_qos_max_parallelism` to understand the capabilities of system in terms of max parallelism, which contains some logic that is like missed here (or at least repeated). I think that if we could figure out a way to get some help from GCD for this purpose, we'd end up with a better (at least in terms of less repeated code, if not in terms of better performance) result.

Reply:
> Shouldn't this be the responsibility of libdispatch?

Johannes suggested on the list that this may not be a good idea because the scheduling is application specific and GCD doesn't know about the application needs (I couldn't fully follow his arguments to be honest). My assumption is/was generally the same, it should be GCD which is dealing with it. It just doesn't appear to work well in the real world? No idea :-)

Three examples I can think of:
- on a device you probably just want to run on a single CPU. (though creating the queue as .background may have the same effect)
- on the server you may want to run a specific number of threads if other servers are assigned to different CPUs (i.e. you want to share load and not grab everything available)
- on the server you may want to run more than a single thread per CPU if you have handlers which are not perfectly async

I find the discussion very interesting, but would like to suggest to move it to the mailing list.

FWIW I just noticed that I may have had a flaw in my thinking that you need serial queues to base serial queues upon. Presumably you can also just attach the connection serials queues on a single concurrent target queue just fine.

Reply:
I created an issue for this in my fork: https://github.com/ZeeZide/http/issues/12

Reply:
just to comment here: In a well behave system I'd say we can expect to be IO bound here, therefore more threads won't help us to be faster. Also there is a latency problem in GCD on Linux when spawning threads because of this: https://github.com/apple/swift-corelibs-libdispatch/blob/666df60ce98b9f38814b544f9ff1ca74b7bbf70d/src/queue.c#L5669 . This however only seems to affect _concurrent_ queues (at least these days).

I wrote a new latency test to assess the situation on Swift 4 with _serial_ queues (see below) and the results are that on Swift 4 with _serial queues_ there are no longer any latency issues. In other words: we can change this code to just use one serial queue per request that isn't targeted at a base queue and it should still work fine.

Without the base queues, I'd expect performance under heavy IO load to be worse as we'd end up with too many threads. There is however a (potentially big) benefit: If a client of this API decides to block (the kernel thread) in a request handler it would no longer an issue. 

I'm happy to switch to one serial queue per request as that would make the code simpler and if someone wants the very best performance it'd be an easy thing to bring back.

```
import Dispatch
import Glibc

let g = DispatchGroup()
g.enter()

func foo(remaining: Int) {
    let q = DispatchQueue(label: "some q")
    if remaining > 0 {
        q.async {
            print("\(pthread_self())")
            foo(remaining: remaining - 1)
        }
        usleep(10_000)
    } else {
        g.leave()
    }
}

foo(remaining: 10000)
g.wait()
```

Reply:
> more threads won't help us to be faster

I think keeping the base queues is worth it for the reverse. I also do on-device servers and I would usually prefer to run those on less threads than the device has cores (one or maybe three on the iPhone Y). The same may be true on servers which are not dedicated to just one app.

What do you think?

Reply:
Also ```usleep(10_000)``` vs ```foo(remaining: 10000)``` 🤔

Reply:
@helje5 the `usleep` is just a crappy way to force it to not re-use the very same thread.

If we want to over-engineer it we could also just have a `baseQueueProvider` closure or something that either returns an appropriate base queue or if the user doesn't want that just `nil`. Because base queue `nil` means no target queue AFAIK.

Reply:
Man, this was 10_000 vs 10000 ;-)

The closure doesn't sound too bad. Would be optional and part of the Options object passed in.

Reply:
I added the closure and dropped the specific algorithm.